tests/proptest.rs

```diff
@@ -0,0 +1,436 @@
+            for _ in 0..3 {
+                match (&one, &resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver)) {
+                    (Ok(l), Ok(r)) => assert_eq!(l, r),
+                    (Err(PubGrubError::NoSolution(derivation_l)), Err(PubGrubError::NoSolution(derivation_r))) => {
+                        prop_assert_eq!(
+                            format!("{}", DefaultStringReporter::report(&derivation_l)),
+                            format!("{}", DefaultStringReporter::report(&derivation_r))
+                        )},
+                    _ => panic!("not the same result")
+                }
+            }
+        }
+    }
+
+    #[test]
+    /// MinimalDependencyProvider change what order the candidates
+    /// are tried but not the existence of a solution
+    fn prop_minimum_version_errors_the_same(
+        (dependency_provider, cases) in registry_strategy(0u16..665, 666)
+    )  {
+        let minimal_provider = MinimalDependencyProvider(dependency_provider.clone());
+        for (name, ver) in cases {
+            let l = resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver);
+            let r = resolve(&TimeoutDependencyProvider::new(minimal_provider.clone(), 50_000), name, ver);
+            match (&l, &r) {
+                (Ok(_), Ok(_)) => (),
+                (Err(_), Err(_)) => (),
+                _ => panic!("not the same result")
+            }
+        }
+    }
+
+    #[test]
+    fn prop_removing_a_dep_cant_break(
+        (dependency_provider, cases) in registry_strategy(0u16..665, 666),
+        indexes_to_remove in prop::collection::vec((any::<prop::sample::Index>(), any::<prop::sample::Index>(), any::<prop::sample::Index>()), 1..10)
+    )  {
+        let packages: Vec<_> = dependency_provider.packages().collect();
+        let mut removed_provider = dependency_provider.clone();
+        for (package_idx, version_idx, dep_idx) in indexes_to_remove {
+            let package = package_idx.get(&packages);
+            let versions = dependency_provider.list_available_versions(package).unwrap();
+            let version = version_idx.get(&versions);
+            let dependencys: Vec<_> = dependency_provider.get_dependencies(package, version).unwrap().unwrap().into_iter().collect();
+            if !dependencys.is_empty() {
+                let dependency = dep_idx.get(&dependencys).0.clone();
+                removed_provider.add_dependencies(
+                    **package,
+                    *version,
+                    dependencys.into_iter().filter(|x| x.0 != dependency)
+                )
+            }
+        }
+        for (name, ver) in cases {
+            if resolve(&dependency_provider, name, ver).is_ok() {
+                prop_assert!(
+                    resolve(&TimeoutDependencyProvider::new(dependency_provider.clone(), 50_000), name, ver).is_ok(),
```

Comment:
Should removing dependencies also ensure shorter runs?

Reply:
That is a really good thought. That is a thing I did not think to test when working on Cargo as it is not flexible like that. Do you think we should work on it in this PR or in a follow up?

Reply:
Nah, jut a thought I had but it's not correct. A missing dependency could generate an incompatibility that later reduces potential packages. But we should check that the difference, if there is one, is not due to timeouts I think.